scipy/integrate/_tanhsinh.py

```diff
@@ -0,0 +1,755 @@
+# mypy: disable-error-code="attr-defined"
+import numpy as np
+from scipy import special
+from scipy.optimize import OptimizeResult
+from scipy.optimize._zeros_py import (_scalar_optimization_initialize,
+                                      _scalar_optimization_loop,
+                                      _ECONVERGED, _ESIGNERR, _ECONVERR,  # noqa
+                                      _EVALUEERR, _ECALLBACK, _EINPROGRESS)  # noqa
+
+# todo:
+#  figure out warning situation
+#  address https://github.com/scipy/scipy/pull/18650#discussion_r1233032521
+#  without `minweight`, we are also suppressing infinities within the interval.
+#    Is that OK? If so, we can probably get rid of `status=3`.
+#  Add heuristic to stop when improvement is too slow / antithrashing
+#  support singularities? interval subdivision? this feature will be added
+#    eventually, but do we adjust the interface now?
+#  When doing log-integration, should the tolerances control the error of the
+#    log-integral or the error of the integral?  The trouble is that `log`
+#    inherently looses some precision so it may not be possible to refine
+#    the integral further. Example: 7th moment of stats.f(15, 20)
```

Comment:
Another instance is the log-integration example in the [top post](https://github.com/scipy/scipy/pull/18650#issue-1749381232). That actually fails to detect convergence, even though it seems to be accurate to the last decimal place.

Reply:
Yeah I've started to think that, too. One thing though is that in https://github.com/scipy/scipy/pull/18050 we made `qmc_quad` return the log of the standard error instead of the standard error of the log.

However, we never actually stated what the behavior of the `error` attribute was in the log case, so we could change it.